woocommerce/woocommerce-ios
# Auth Role Checks: Part 1 - Add User-related networking and store classes

## Summary

Refs #3963 

This is the first out of three parts that make up the changes related to role authentication (for more details, refer to #3963). Also, note that this PR targets `issue/3963-auth-role-error` branch. 

Here's a summary of this change:
- Added a new `User` model, with its own `Remote`, `Mapper`, `Store`, and `Action`. The `User` model represents a site-specific user information (e.g. `roles` being the most important one). As for the name, I previously considered `SiteUser` and `SiteAccount`, but decided to go ahead with `User` for now. Let me know if you prefer otherwise!
- The user information is fetched via WP Core v2 API, through this endpoint: `/sites/$site_id/users/me`. It's also important that the request includes `{"context": "edit"}`, since the `roles` property is only visible through that context.
- The key information to fetch from the `User` model is the `roles`, and this is why the model is not persisted in the local database – there's no use case for it as of now. We want to make sure that only users with `administrator` or `shop_manager` roles can get into the main part of the app. 
  - In the later pull requests, this information will be stored locally in user defaults in a form of boolean, just to cache the last check result.
  - Ideally, we almost always want the latest state to ensure that the user doesn't stay blocked for too long after their role has been updated by those in charge.

## To Test

There's no UI or flow to test yet, so:

- Run unit tests on Networking module, and make sure everything is 🆗  .
- Run tests on Yosemite module, and make sure everything is 🆗 .

## Author Checklist 
Update release notes:
- [x] Add unit tests where applicable.
